File: benchmarks/JavaScript/permute.js
// @ts-check
// This code is derived from the SOM benchmarks, see AUTHORS.md file.
//
// Copyright (c) 2015-2016 Stefan Marr <git@stefan-marr.de>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the 'Software'), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

const { Benchmark } = require('./benchmark');

class Permute extends Benchmark {
  constructor() {
    super();
    this.count = 0;
    this.v = null;
  }

  benchmark() {
    this.count = 0;
    this.v = new Array(6).fill(0);
    this.permute(6);
    return this.count;
  }

  verifyResult(result) {
    return result === 8660;
  }

  permute(n) {
    this.count += 1;
    if (n !== 0) {
      const n1 = n - 1;
      this.permute(n1);
      for (let i = n1; i >= 0; i -= 1) {
        this.swap(n1, i);
        this.permute(n1);
        this.swap(n1, i);
      }
    }
  }

  swap(i, j) {
    const tmp = this.v[i];
    this.v[i] = this.v[j];
    this.v[j] = tmp;
  }
}

exports.newInstance = () => new Permute();

File: benchmarks/CSharp/Permute.cs
using System;

public class Permute : Benchmark
{
    int count;
    int[] v;

    public static void Main(string[] args){
        Permute s = new Permute();
        Object result = s.benchmark();
        bool rr = s.verifyResult(result);
    }

    public override Object benchmark(){
        count = 0;
        v = new int[6];
        permute(6);
        return count;
    }

    public void permute(int n){
        count++;
        if(n !=0){
            int n1 = n -1;
            permute(n1);
            for(int i = n1; i >=0; i--){
                swap(n1,i);
                permute(n1);
                swap(n1, i);
            }
        }
    }

    public void swap(int i, int j){
        int tmp = v[i];
        v[i] = v[j];
        v[j] = tmp;
    }

    public override bool verifyResult(Object result){
        return (int) result == 8660;
    }
}